tiledb/sm/filesystem/posix.cc

```diff
@@ -299,6 +300,56 @@ tuple<Status, optional<std::vector<directory_entry>>> Posix::ls_with_sizes(
   return {Status::Ok(), entries};
 }
 
+bool Posix::ls_cb(
+    const URI& uri, LsCallback cb, void* data, bool recursive) const {
+  std::string path = uri.to_path();
+  struct dirent** paths;
+  // scandir to traverse sorted objects, so we can invoke callback directly.
+  int path_num = scandir(path.c_str(), &paths, NULL, alphasort);
```

Comment:
A raw call to a function (in this case a system call) that directly allocates memory needs to be wrapped in exception-safe types at the point of call, not afterwards, as is done here. `scandir` allocates ones for each string, which means that we need an exception-safe string class. Nothing in the standard library does this, so we'll have to write one. The absence of such wrapper shows up as raw calls to `free` here and some other ones in test code.